src/events/guildMemberRemove.ts

```diff
@@ -0,0 +1,113 @@
+
+const logger = getLogger("guildMemberRemove");
+
+type Tail<T extends unknown[]> = T extends [unknown, ...infer R] ? R : never;
+
+/**
+ * Helper function to send the same message to the list of channels and
+ * send any exceptions to the logger.
+ * @nothrow
+ */
+async function messageChannels(
+	discord: DiscordInterface,
+	channels: string[],
+	...args: Tail<Parameters<DiscordInterface["sendMessage"]>>
+): Promise<void> {
+	for (const channel of channels) {
+		await discord.sendMessage(channel, ...args).catch(logger.error);
+	}
+}
+
+export function makeHandler(database: DatabaseWrapperPostgres, discord: DiscordInterface, website: WebsiteInterface) {
+	return async function guildMemberRemove(server: Guild, member: Member | MemberPartial): Promise<void> {
+		if (member.user.bot) {
+			return;
+		}
+		// Remove Participant from every preparing or in progress tournament in the database
+		const dropped = await database.dropFromAll(server.id, member.id).catch(logger.error);
+		if (!dropped || !dropped.length) {
+			return;
+		}
+		logger.verbose(
+			JSON.stringify({
+				server: server.id,
+				name: server.name,
+				id: member.id,
+				username: `${member.user.username}#${member.user.discriminator}`,
+				tournaments: dropped.map(({ tournamentId, challongeId }) => ({ tournamentId, challongeId }))
+			})
+		);
+		const who = `<@${member.id}> (${member.user.username}#${member.user.discriminator})`;
+		for (const { tournamentId, privateChannels, challongeId } of dropped) {
+			// For each tournament, inform the private channel that the user left and was dropped.
+			await messageChannels(
+				discord,
+				privateChannels,
+				`${who} dropped from **${tournamentId}** by leaving the server.`
+			);
+			// For each tournament in progress, inform the opponent that their opponent dropped
+			if (challongeId !== undefined) {
+				// modified from TournamentManager.sendDropMessage
+				const match = await website.findMatch(tournamentId, challongeId);
+				// if there's no match, their most recent score is already submitted.
+				if (match) {
+					const oppChallonge = match.player1 === challongeId ? match.player2 : match.player1;
+					try {
+						await website.submitScore(tournamentId, match, oppChallonge, 2, 0);
+						logger.verbose(
+							JSON.stringify({
+								id: member.id,
+								tournament: tournamentId,
+								match: match.matchId,
+								oppChallonge
+							})
+						);
+						const { discordId } = await website.getPlayer(tournamentId, oppChallonge);
+						logger.verbose(
+							JSON.stringify({
+								id: member.id,
+								tournament: tournamentId,
+								match: match.matchId,
+								oppChallonge,
+								discordId
+							})
+						);
+						if (discordId !== "DUMMY") {
+							await discord.sendDirectMessage(
+								discordId,
+								`Your opponent ${who} has dropped from the tournament, conceding this round to you. You don't need to submit a score for this round.`
+							);
+						}
```

Comment:
I'm not sure if more should be done. The cases when `misc === null` on Challonge should be:
- a dummy bye, which automatically is removed when the round starts, so the guildMemberRemove event would have to interleave with that
- users added from the Challonge side? I don't think this will be an external use case moving forward

Reply:
If it's a dummy bye, the match should be closed before a player can leave, right? The only exception I can think of is due to the slow operation of the Challonge API when a tournament is starting, but theoretically in that case the request for match information should be delayed the same way and so still not return until the match is closed.

Reply:
I'm wondering if the check is necessary, if more checks are required to avoid possible extra errors logged, and if anything should be logged in the else case. The "interleaving" situation I recognize is possible but will leave it to the far future to address since that'll require mutexes for each tournament (I know, mutexes in a single-threaded program)